internal/database/types_controller.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2025 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Azure/ARO-HCP/internal/api"
+	azcorearm "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+)
+
+type Controller struct {
+	TypedDocument `json:",inline"`
+
+	ControllerProperties ControllerProperties `json:"properties"`
+}
+
+var _ ResourceProperties = &Controller{}
+
+type ControllerProperties struct {
+	// ExternalID is the Azure resource ID of the type this is associated with.
+	ExternalID *azcorearm.ResourceID `json:"externalId,omitempty"`
```

Comment:
In what cases would the resource ID be empty?

Reply:
Shouldn't ever be empty, IIUC.  We maybe overuse the `omitempty` JSON tag.